lib/c-api/src/wasm_c_api/instance.rs

```diff
@@ -32,7 +31,35 @@ pub unsafe extern "C" fn wasm_instance_new(
         .cloned()
         .collect();
 
-    let instance = Arc::new(c_try!(Instance::new(wasm_module, &resolver)));
+    let instance = match Instance::new(wasm_module, &resolver) {
+        Ok(instance) => Arc::new(instance),
+
+        Err(InstantiationError::Link(link_error)) => {
+            crate::error::update_last_error(link_error);
+
+            return None;
+        }
+
+        Err(InstantiationError::Start(runtime_error)) => {
+            let pointer = {
+                let trap: Box<wasm_trap_t> = Box::new(runtime_error.into());
+                let mut traps: Vec<*mut wasm_trap_t> = Vec::with_capacity(1);
+                traps.push(Box::into_raw(trap));
+
+                debug_assert!(traps.len() == traps.capacity());
```

Comment:
Do we really need this here?

Reply:
I think it's probably good to have, alternatively we can call `shrink_to_fit` on the vector and guarantee that it's always true.  Methods like `into_boxed_slice` will shrink the allocation too, because slices have no notion of capacity

Reply:
I believe it would be good to not have it here.

The debug asserts have bitten us before, and the logic is implied by `Vec::with_capacity(1)`

Reply:
Removed.

Reply:
To be fair, the debug asserts did exactly what they were supposed to: they caught a memory leak 😆 the primary issue is that we don't test with debug_asserts enabled so we didn't catch the bug, a user did... perhaps they should just be normal `assert!` or we should test with `debug_assert!`...

Reply:
That said, I strongly prefer writing code in such a way that this isn't necessary, we were able to abstract a lot of our Vec into pointer code away and now this logic can't be messed up.